src/cpu/flexcpu/simple_dataflow_thread.hh

```diff
@@ -0,0 +1,620 @@
+#include "cpu/inst_seq.hh"
+#include "cpu/reg_class.hh"
+#include "cpu/simple_thread.hh"
+#include "mem/request.hh"
+
+class SimpleDataflowCPU;
+
+/**
+ * Instances of this class represent a logical "Hyperthread" within the
+ * SimpleDataflowCPU. A thread contains an independent complete architectural
+ * state, and in this case, also the logical control flow.
+ */
+class SDCPUThread : public ThreadContext
+{
+  protected:
+    // BEGIN SDCPUThread Internal definitions
+
+    class MemIface : public InflightInst::MemIface
+    {
+      protected:
+        SDCPUThread& sdCPUThread;
+      public:
+        MemIface(SDCPUThread& sdCPUThread_):
+            sdCPUThread(sdCPUThread_)
+        { }
+
+        Fault initiateMemRead(std::shared_ptr<InflightInst> inst, Addr addr,
+                             unsigned int size, Request::Flags flags) override;
+
+        Fault writeMem(std::shared_ptr<InflightInst> inst, uint8_t *data,
+                       unsigned int size, Addr addr, Request::Flags flags,
+                       uint64_t *res) override;
+    };
+
+    struct SplitRequest {
+        RequestPtr main = nullptr;
+        RequestPtr low = nullptr;
+        RequestPtr high = nullptr;
+    };
+
+    // END SDCPUThread Internal definitions
+
+    // BEGIN SDCPUThread Internal variables
+
+    /**
+     * We hold a pointer to the CPU which owns this logical thread.
+     */
+    SimpleDataflowCPU* _cpuPtr;
+
+    /**
+     * We store a name for use with tracing
+     */
+    std::string _name;
+
+    /**
+     * We implement the memory interface of the InflightInst so that we can
+     * capture calls to request memory access via our CPU.
+     */
+    MemIface memIface;
+
+    /**
+     * The SimpleThread class already contains all internal state necessary to
+     * hold onto the state of a basic CPU, demonstrated by its use in many
+     * other CPU models. Until we find a significant need to hold particular
+     * permanent committed state different from a regular CPU in this library,
+     * there is no reason to redefine the wheel.
+     *
+     * The committed state represents the state resulting from the complete
+     * correct execution of all instructions up to the point represented by
+     * numInstsCommitted.
+     */
+    SimpleThread* _committedState;
+
+    /**
+     * We hold an ISA pointer of our own because _committedState doesn't expose
+     * its.
+     */
+    TheISA::ISA* isa;
+
+    /**
+     * We hold onto the InstSeqNum of the last instruction successfully
+     * committed, for ease of checking if older instruction dependencies have
+     * already been satisfied.
+     */
+    InstSeqNum lastCommittedInstNum = 0;
+
+    // BEGIN Speculative state
+
+    /**
+     * For use when we need to fetch multiple MachInsts for one decode.
+     */
+    Addr fetchOffset = 0;
+
+    /**
+     * Number of fetch actions we have begun and not finished at this time
+     */
+    size_t outstandingFetches = 0;
```

Comment:
Why needed? It's not used in the code right now. If this is for constraining the fetch bandwidth, I think that should be done by the CPU, not the thread.